### PR TITLE
added all java files

### PR DESCRIPTION
**Pull request by TimSon1022**

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,33 +1,122 @@
...
+        if(startWord <= 0 || endWord <= 0) {
+        	throw new IllegalArgumentException("Error: start index and end index is less than 1");
+        }
+
+        else if (startWord > endWord) {
+            throw new IndexOutOfBoundsException("Error: start index is larger than end");
+        }
+
+        else if (endWord > words.length) {
+            throw new IndexOutOfBoundsException("Error: end index is larger than the string length");
```
Comment (chiouc) [inline src/main/java/StringManipulation.java:84]: I think you would be able to combine this else if statement with the one from line 79. This will help make the code shorter and will combine the same type of exception. 
Ex: else if (startWord > endWord || endWord > words.length) {
           throw new IndexOutOfBoundsException(); 
      }

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,33 +1,122 @@
...
+
+
+
+
 
     @Override
     public String[] getSubStrings(int startWord, int endWord) {
-        return null;
+        if (string == null) {
+        	throw new NullPointerException("Error: string is not yet initialized");
```
Comment (chiouc) [inline src/main/java/StringManipulation.java:72]: Null Pointer Exception error isn't required for this method. I'm pretty sure this is actually supposed to be  an Illegal Argument Exception so I would modify this as part of that exception.

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,33 +1,122 @@
...
+    	if (string == null) {
+            throw new NullPointerException("Error: string is not yet initialized");
+        }
 
-
-}
+        if (indices.length != string.length()) {
+            throw new IllegalArgumentException("Error: string length is not equal to indices length");
+        }
+        for (int i = 0; i < indices.length - 1; i++) {
```
Comment (chiouc) [inline src/main/java/StringManipulation.java:107]: This outer nested for loop seems redundant. I'm not sure why you need to run the for loop for indices.length - 1. I believe you can just combine the 2 for loops into for (int i = 0; i < indices.length; i++) and then just run the exception in your if statement for if indices[i]< 0  or  indices[i]>= string length

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,33 +1,122 @@
...
+                if (indices[i] == indices[j]) {
+                    throw new IllegalArgumentException("Error: some indexes are not unique");
+                }
+            }
+        }
+        
+        char[] restored = new char[indices.length];
+        for (int i = 0; i < indices.length; i++) {
+        	
+            restored[i] = string.charAt(indices[i]);
```
Comment (chiouc) [inline src/main/java/StringManipulation.java:118]: you can run this code in the original for loop at line 108 instead of writing a whole new for loop for this portion.

File: src/test/java/StringManipulationTest.java [hunk 5]
```diff
@@ -24,90 +24,159 @@ public void testCount1() {
...
         assertEquals(sStings[1], "string");
     }
+ // this test checks whether method getSubStrings returns the right substring, which is "my" and "string"
 
     @Test
-    public void testGeSubStrings2() {
-        fail("Not yet implemented");
+    public void testGetSubStrings2() {
+    	manipulatedstring.setString(null);
+        NullPointerException testException = assertThrows(NullPointerException.class, () -> {
```
Comment (chiouc) [inline src/test/java/StringManipulationTest.java:129]: I don't think you need a test for null pointer exception since this would fall under the other exceptions. It might be best to redo this test so you can test this under illegal argument exception

**Comment by chiouc** [comment]
Hi Tim, your code overall looks functional, concise, and is all very readable. I think most of the tests (besides the null pointer exception tests) test all the edge cases. Since the code works, I would just recommend some refactoring for cleaner code.

**Review by chiouc — APPROVED** [review]
